### PR TITLE
Make OAuth errors explicit

### LINKED ISSUE #1232: Stack returns Error 500 on authentication
#### Summary

I am setting up a private network using the TTN V3 Stack on a k8s local server.
The whole setup process goes fine, until the CLI authentication.
After entering login/pwd on the web UI i get a 500 Error page.
Attached, you will find the logs with debug level activated.

...

#### Steps to Reproduce

 1 - Get a minikube or a full k8s cluster
2. - Use the attached kubernetes files to setup the infrastructure 
[ttn-k8s.tar.gz](https://github.com/TheThingsNetwork/lorawan-stack/files/3534242/ttn-k8s.tar.gz)
3. - log with the CLI through the web, enter the credentials 

#### What is already there? What do you see now?

logs with debug activated doesn't really help but here they are.
[logs-from-stack-in-stack-dc4c979b5-4zhzq.txt](https://github.com/TheThingsNetwork/lorawan-stack/files/3534202/logs-from-stack-in-stack-dc4c979b5-4zhzq.txt)

Is there any way to get a more verbose logging ?
...

#### What is missing? What do you want to see?

I want to get my oAuth Token, i get a 500 error
...

#### Environment

OS : Ubuntu 16
K8s : V1.15.2
...

#### How do you propose to implement this?

...

#### Can you do this yourself and submit a Pull Request?

...

### PR DESCRIPTION
**Pull request by htdvisser**
#### Summary

This pull request makes OAuth errors explicit by turning error IDs returned by the OAuth library into defined errors.

Closes #1232 and perhaps also #1226.

#### Changes

- Return defined errors from `/oauth/authorize` if possible
- Inspect errors from OAuth library and convert to defined errors

#### Release Notes

- Improved error handling in OAuth flow

### CODE REVIEW

File: pkg/oauth/oauth.go [hunk 1]
```diff
@@ -73,24 +80,16 @@ func (s *server) Authorize(authorizePage echo.HandlerFunc) echo.HandlerFunc {
 		ar.UserData = userData{UserIdentifiers: session.UserIdentifiers}
 		client := ttnpb.Client(ar.Client.(osinClient))
 		if !clientHasGrant(&client, ttnpb.GRANT_AUTHORIZATION_CODE) {
-			resp.SetError(osin.E_INVALID_CLIENT, "OAuth client does not have authorization code grant")
-			oauth2.FinishAuthorizeRequest(resp, req, ar)
-			return s.output(c, resp)
+			return errClientMissingGrant.WithAttributes("grant", "authorization_code")
```
Comment (johanstokking) [inline pkg/oauth/oauth.go]: How is `E_INVALID_CLIENT` propagated to the caller?

Comment (htdvisser) [reply pkg/oauth/oauth.go]: Previously these errors just resulted in a 500 error (the issue). Now we return an explicit error in the same way as all other errors. The error is rendered into the HTML page and then picked up by the webui code.

Comment (johanstokking) [reply pkg/oauth/oauth.go]: OK I see, but with this explicit call to `SetError()` with a code I was under the impression it's some application layer transport of error codes, i.e. as part of OAuth and not just HTTP error codes

Comment (htdvisser) [reply pkg/oauth/oauth.go]: Not for these internal errors. Some errors are indeed handled by the OAuth protocol (with an error param in the redirect), but if the OAuth client is invalid, we shouldn't redirect anyway.

Comment (htdvisser) [reply pkg/oauth/oauth.go]: Ok, it turns out that the OAuth spec in fact allows redirecting on these errors, I'll revert the deletes.

**Pull request merged**